DM-29011: Add support for ButlerURI.walk()

timj opened the pull request:
Works for S3 and File URIs.

timj committed:
Add ButlerURI.walk and findFileResources to File URIs

timj committed:
Add grouped mode for finding files and grouping by directory

timj committed:
Add ButlerURI.walk() for S3

timj committed:
Document that uri.join("/") always returns a file:/// URI

It's arguable that this was never the intent and should always
complain.

timj committed:
Ensure that uri.join() for a dirLike thing looks like a dir URI

normpath strips the trailing / and we need to add it back otherwise
geturl does not include it.

timj committed:
walk() can return [] so trap that

timj committed:
Add tests for s3.walk

timj committed:
Fix mypy for walk signature

os.walk can return [] so mypy has to be told this is possible.

ktlim commented on python/lsst/daf/butler/core/_butlerUri/s3.py:
```diff
@@ -226,3 +229,75 @@ def transfer_from(self, src: ButlerURI, transfer: str = "copy",
+        s3_paginator = self.client.get_paginator('list_objects_v2')
+
+        # Limit each query to a single "directory" to match os.walk
+        # We could download all keys at once with no delimiter and work
```
Because it's paginated, if you're going to scan them all eventually anyway, I think you'd be better off doing the simple version — except that you then wouldn't have the same result ordering as `os.walk()`.

ktlim commented on tests/test_uri.py:
```diff
@@ -343,6 +395,40 @@ def testTransfer(self):
+    def testWalk(self):
+        """Test that we can list an S3 bucket"""
+        # Files we want to create
+        expected = ("a/x.txt", "a/y.txt", "a/z.json", "a/b/w.txt", "a/b/c/d/v.json")
```
I think you should have at least one test that extends over more than one page, which unfortunately means a lot of files, given the default max of 1000 keys.

ktlim commented on python/lsst/daf/butler/core/_butlerUri/_butlerUri.py:
```diff
@@ -771,3 +788,81 @@ def transfer_from(self, src: ButlerURI, transfer: str,
+            The regex to use when searching for files within directories.
+            By default returns all the found files.
+        grouped : `bool`, optional
+            If `True` the results will be grouped by directly and each
```
```suggestion
            If `True` the results will be grouped by directory and each
```

timj committed:
Change grouping for findFileResources to group all explicit files

Returning a lot of individual 1-file groups is not going to be
useful for raw ingest. Makes more sense to combine them
all into a single group.

timj committed:
Update comment on s3.walk

Explains that we are not downloading all the keys in one
go and collating locally because that can become an unbounded
problem for large buckets.

timj committed:
Test that pagination is working for large directories
